### PR TITLE
Does the versioning stuff, using the Puts

### PR DESCRIPTION
**Pull request by dakota002**
This replaces #16  and #17

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -96,15 +114,23 @@ export class DynamoDBAutoIncrement {
           Item: {
             ...this.props.counterTableKey,
             [this.props.counterTableAttributeName]: nextCounter,
+            ...lastItem,
```
Comment (lpsinger) [inline src/index.ts:117]: No, you want to update this with `item`, not `lastItem`.

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 1 above)
Comment (dakota002) [reply src/index.ts:117]: No, in this case I want to save the last item. The previous version for the item being updated will get saved in the PUT to `this.props.counterTableName` (lines 109-120) , and the new version with the updated counter value gets saved in the PUT to `this.props.tableName` (121-135)

Comment (lpsinger) [reply src/index.ts:117]: Don't we want the _current_ version and _current_ item contents saved to both tables?

File: src/index.ts [hunk 2]
```diff
@@ -64,7 +67,9 @@ export class DynamoDBAutoIncrement {
         await this.props.doc.get({
           AttributesToGet: [this.props.counterTableAttributeName],
           Key: this.props.counterTableKey,
-          TableName: this.props.counterTableName,
+          TableName: this.props.counterTableCopyItem
+            ? this.props.tableName
+            : this.props.counterTableName,
```
[the remark this answers is not in the document]
Comment (dakota002) [reply src/index.ts:72]: For the history tables that would contain the item copies, I don't think it makes sense to store the Latest version in the history table, since it would be a duplicate of the data in the main table. 

For circulars, the idea being that the `circulars` table has some item:
```
{
  'circularId': 12345,
  'version': 2,
  'subject': ...,
  'body': ...
}
```
This is the latest version, and is about to be updated to:
```
{
  'circularId': 12345,
  'version': {# to be determined},
  'subject': ...,
  'body': ...
}
```

The history table only contains the old versions, so just:
```
{
  'circularId': 12345,
  'version': 1,
  'subject': ...,
  'body': ...
}
``` 
So the latest value is the `version` attribute from the `circulars` table

[the remark this answers is not in the document]
Comment (lpsinger) [reply src/index.ts:72]: I need to think about this a bit. Storing all versions including the latest in the history table has the advantage that it will simplify querying the versions of a document: only one query is required. Storing all versions _except_ the latest in the history table has the advantage of requiring half as much storage, in the limit that most documents have only one version.

File: src/index.ts (hunk 1 above)
Comment (dakota002) [reply src/index.ts:117]: No, I don't think it makes sense to save the data twice, right? Queries against the history table should only be retrieving old versions of the data. This was at least the original assumption made in the demo version.

File: src/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dakota002) [reply src/index.ts:72]: Yea, it seems our comments in this and below are heading to the same point. I had settled on the latter for this design

File: src/index.ts (hunk 1 above)
Comment (lpsinger) [reply src/index.ts:117]: I prefer to store the current version in both tables. It simplifies the access pattern for retrieving a document by version number because it requires only one request instead of possibly two. It also makes this a one-line change. That's what I had in mind with #19.

Comment (dakota002) [reply src/index.ts:117]: Okay that should be an easier implementation, I'll updated the PR

**Comment by lpsinger** [comment]
@dakota002, I've come around to your original idea. It has many advantages over mine:

 - It uses half as much storage space in DynamoDB.
 - It will require no database migration.
 - In both the Circular auto increment application and the version numbering application, only two tables are involved, not three.

However it does add a bunch of conditional logic, which is a hint that we should be pulling out the common logic and reusing it in two new classes. I am going to propose a couple refactorings in separate PRs that will make this more elegant.

**Comment by lpsinger** [comment]
At this point, @dakota002, please implement your original versioning algorithm, but by subclassing the base class.

**Review by lpsinger — CHANGES_REQUESTED** [review]
What I'm looking for is a new class that extends `BaseDynamoDBAutoIncrement` and implements the `next` method. You shouldn't need to modify the existing `DynamoDBAutoIncrement` class, and you shouldn't need to add a flag to the props.

**Comment by dakota002** [comment]
> What I'm looking for is a new class that extends `BaseDynamoDBAutoIncrement` and implements the `next` method. You shouldn't need to modify the existing `DynamoDBAutoIncrement` class, and you shouldn't need to add a flag to the props.

Sounds good!

File: src/index.ts [hunk 3]
```diff
@@ -102,6 +101,7 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
     const counter = await this.#getLast()
 
     let nextCounter, ConditionExpression, ExpressionAttributeValues
+
```
Comment (dakota002) [inline src/index.ts:104]:
```suggestion
```

File: src/index.ts [hunk 4]
```diff
@@ -35,7 +35,6 @@ abstract class BaseDynamoDBAutoIncrement {
   async put(item: Record<string, NativeAttributeValue>) {
     for (;;) {
       const { puts, nextCounter } = await this.next(item)
-
```
Comment (lpsinger) [inline src/index.ts:38]: Undo unrelated whitespace change.

File: src/index.ts [hunk 5]
```diff
@@ -139,3 +138,110 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+ * import { DynamoDB } from '@aws-sdk/client-dynamodb'
+ * import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb'
+ * import { DynamoDBHistoryAutoIncrement } from '@nasa-gcn/dynamodb-autoincrement'
+ *
+ * const client = new DynamoDB({})
+ * const doc = DynamoDBDocument.from(client)
+ *
+ * const autoIncrementHistory = DynamoDBHistoryAutoIncrement({
+ *   doc,
+ *   counterTableName: 'autoincrementHistoryTable',
```
Comment (lpsinger) [inline src/index.ts:156]:
```suggestion
 *   counterTableName: 'widgetsHistory',
```

File: src/index.ts [hunk 6]
```diff
@@ -139,3 +138,110 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+ *   counterTableKey: {
+ *     autoincrementItemID: 1
```
Comment (lpsinger) [inline src/index.ts:158]: Pick an example widget ID other than 1 to make it clear that this is _not_ an initial value, but a primary key value:

```suggestion
 *     widgetID: 42
```

File: src/index.ts [hunk 7]
```diff
@@ -139,3 +138,110 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+ *   },
+ *   attributeName: 'version',
+ *   tableName: 'widgets',
+ *   initialValue: 1,
+ * })
+ *
+ * const latestVersionValue = await autoIncrementHistory.put({
```
Comment (lpsinger) [inline src/index.ts:165]:
```suggestion
 * const latestVersionNumber = await autoIncrementHistory.put({
```

File: src/index.ts [hunk 8]
```diff
@@ -139,3 +138,110 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+    const counter = await this.#getLast()
+
+    let nextCounter
+
+    const existingUntrackedEntry = (
+      await this.props.doc.get({
+        TableName: this.props.tableName,
+        Key: this.props.counterTableKey,
+      })
+    ).Item
```
Comment (lpsinger) [inline src/index.ts:194]: No need to get the last item twice...

```suggestion
  protected async next(item: Record<string, NativeAttributeValue>) {
    let nextCounter

    const existingUntrackedEntry = (
      await this.props.doc.get({
        TableName: this.props.tableName,
        Key: this.props.counterTableKey,
      })
    ).Item

    const counter: number | undefined = existingUntrackedEntry?.[this.props.attributeName]
```

File: src/index.ts [hunk 9]
```diff
@@ -139,3 +138,110 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+
+    let untractedEntryPutCommandInput: PutCommandInput | undefined = undefined
```
Comment (lpsinger) [inline src/index.ts:196]: Untracted? Did you mean untracked?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 9 above)
Comment (dakota002) [reply src/index.ts]: Yup! Fixed in latest commit

File: src/index.ts [hunk 10]
```diff
@@ -139,3 +139,99 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+ *   widgetName: 'A new name for this item',
+ *   costDollars: 199.99,
+ * })
+ * ```
+ */
+export class DynamoDBHistoryAutoIncrement extends BaseDynamoDBAutoIncrement {
+  protected async next(item: Record<string, NativeAttributeValue>) {
+    let nextCounter
+
+    const existingUntrackedEntry = (
```
Comment (lpsinger) [inline src/index.ts:176]: It's not necessarily untracked, it's just the existing entry.

```suggestion
    const existingEntry = (
```

File: src/index.ts [hunk 11]
```diff
@@ -139,3 +139,99 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+ * const doc = DynamoDBDocument.from(client)
+ *
+ * const autoIncrementHistory = DynamoDBHistoryAutoIncrement({
+ *   doc,
+ *   counterTableName: 'widgetsHistory',
+ *   counterTableKey: {
+ *     widgetID: 42
+ *   },
+ *   attributeName: 'version',
+ *   tableName: 'widgets',
```
Comment (lpsinger) [inline src/index.ts:162]: For closer similarity to the behavior of the `DynamoDBAutoIncrement` class, the table named `counterTableName` should be the table that stores the current version number, and hence also the current item. The table named `tableName` should be the one that accumulates prior versions. Please switch the names of these two tables in your example.

With that change, do you need to update the implementation as well?

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 11 above)
Comment (dakota002) [reply src/index.ts:162]: Hm likely yes, I'll make some changes locally and see what all gets shifted around

File: src/index.ts (hunk 10 above)
Comment (dakota002) [reply src/index.ts:176]: Will be resolved in next commit

File: src/index.ts [hunk 12]
```diff
@@ -139,3 +139,99 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+          '#counter': this.props.attributeName,
+        },
+        ExpressionAttributeValues: {
+          ':counter': nextCounter,
+        },
+        Item,
+        TableName: this.props.counterTableName,
+      },
+    ]
+    if (untrackedEntryPutCommandInput) puts.push(untrackedEntryPutCommandInput)
```
Comment (lpsinger) [inline src/index.ts:233]: I think we agreed that we would only store the _current_ value in counterTable, and only _past_ values in table.

```suggestion
    const existingItem = (
      await this.props.doc.get({
        TableName: this.props.counterTableName,
        Key: this.props.counterTableKey,
      })
    ).Item

    const counter: number | undefined =
      existingItem?.[this.props.attributeName]

    if (counter === undefined) {
      nextCounter = this.props.initialValue

      // Existing item didn't have a version, so give it one
      if (existingItem) {
        existingItem[this.props.attributeName] = nextCounter
        nextCounter += 1
      }
    } else {
      nextCounter = counter + 1
    }

    const puts: PutCommandInput[] = [
      {
        ConditionExpression: 'attribute_not_exists(#counter)',
        ExpressionAttributeNames: {
          '#counter': this.props.attributeName,
        },
        Item: existingItem,
        TableName: this.props.tableName,
      },
      {
        ConditionExpression:
          'attribute_not_exists(#counter) OR #counter <> :counter',
        ExpressionAttributeNames: {
          '#counter': this.props.attributeName,
        },
        ExpressionAttributeValues: {
          ':counter': nextCounter,
        },
        Item: {
          ...item,
          ...this.props.counterTableKey,
          [this.props.attributeName]: nextCounter,
        },
        TableName: this.props.counterTableName,
      },
    ]
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 12 above)
Comment (dakota002) [reply src/index.ts:233]: Ah yea you are right, I will need to update the tests and make sure this is behaving as expected

File: src/index.ts [hunk 13]
```diff
@@ -139,3 +139,88 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+    let nextCounter
+
+    const existingItem = (
+      await this.props.doc.get({
+        TableName: this.props.counterTableName,
+        Key: this.props.counterTableKey,
+      })
+    ).Item
+
+    const counter: number | undefined = existingItem?.[this.props.attributeName]
```
Comment (lpsinger) [inline src/index.ts:183]:
```suggestion
    let counter: number | undefined = existingItem?.[this.props.attributeName]
```

File: src/index.ts [hunk 14]
```diff
@@ -139,3 +139,88 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+
+    if (counter === undefined) {
+      nextCounter = this.props.initialValue
+
+      // Existing item didn't have a version, so give it one
+      if (existingItem) {
+        existingItem[this.props.attributeName] = nextCounter
```
Comment (lpsinger) [inline src/index.ts:190]:
```suggestion
        existingItem[this.props.attributeName] = counter = nextCounter
```

File: src/index.ts [hunk 15]
```diff
@@ -139,3 +139,88 @@ export class DynamoDBAutoIncrement extends BaseDynamoDBAutoIncrement {
...
+      },
+      {
+        ConditionExpression:
+          'attribute_not_exists(#counter) OR #counter <> :counter',
+        ExpressionAttributeNames: {
+          '#counter': this.props.attributeName,
+        },
+        ExpressionAttributeValues: {
+          ':counter': nextCounter,
+        },
```
Comment (lpsinger) [inline src/index.ts:214]: Compare with _old_ value of counter:

```suggestion
        ConditionExpression:
          'attribute_not_exists(#counter) OR #counter = :counter',
        ExpressionAttributeNames: {
          '#counter': this.props.attributeName,
        },
        ExpressionAttributeValues: {
          ':counter': counter,
        },
```

File: src/test.ts [hunk 16]
```diff
@@ -43,18 +55,49 @@ afterEach(async () => {
...
+                    Key: {
+                      [PartitionKeyAttributeName]: KeyValue,
+                      [SortKeyAttributeName]: SortKeyValue,
+                    },
+                  })
+              )
+            )
         )
-    )
+      )
```
Comment (lpsinger) [inline src/test.ts:100]: Refactor to reduce code duplication:

```ts
  await Promise.all(
    [
      { TableName: 'autoincrement', keyAttributeNames: ['tableName'] },
      { TableName: 'widgets', keyAttributeNames: ['widgetID'] },
      ... // etc.
    ].map(
      async ({ TableName, keyAttributeNames }) =>
        await Promise.all(
          ((await doc.scan({ TableName })).Items ?? []).map(
            async (item) =>
              await doc.delete({
                TableName,
                Key: Object.fromEntries(
                  keyAttributeNames.map((key) => [key, item[key]])
                ),
              })
          )
        )
    )
  )
```

**Review by lpsinger — CHANGES_REQUESTED** [review]
- [x] Please add a unit test that is analogous to `correctly handles a large number of parallel puts`, that tests that the locking is working correctly.

**Review by lpsinger — APPROVED** [review]
